spec.md

```diff
@@ -1586,12 +1760,192 @@ en <= Z_valid
 cover(clk, pred, en, "X equals Y when Z is valid") : optional_name
 ```
 
+## Probes
+
+Probe references are created with `probe`{.firrtl} expressions, exported with
+the `export`{.firrtl} statement, forwarded between instances using the
+`forward`{.firrtl} statement, read using the `read`{.firrtl} expression (see
+[@sec:reading-probe-references]), and used with `force`{.firrtl} and
+`release`{.firrtl} statements.
+
+Export and forward are used to route references through the design, and may be
+used wherever is most convenient in terms of available identifiers.
+Every sink-flow probe must be the target of exactly one of these statements.
+
+These statements are detailed below.
+
+### Export
+
+The export statement takes a reference expression, which must be a probe expression,
+and exports the probe reference to into a sink-flow static reference target.
+
+```firrtl
+module Refs:
+  input clock:  Clock
+  output a : Probe<{x: UInt, y: UInt}> ; read-only ref. to wire 'p'
+  output b : RWProbe<UInt> ; force-able ref. to node 'q', inferred width.
+  output c : Probe<UInt<1>> ; read-only ref. to register 'r'
+  output d : RWProbe<UInt<2>>[4] ; vector of ref.'s to memory data in 'm'
+  output e : Probe<Clock> ; ref. to input clock port
+
+  wire p : {x: UInt, flip y : UInt}
+  export probe(p) as a ; probe is passive
+  node q = UInt<1>(0)
+  export rwprobe(q) as b
+  reg r: UInt, clock
+  export probe(r) as c
+  mem m:
+    data-type => UInt<5>
+    depth => 4
+    ; ...
+    read-under-write => undefined
+
+  export rwprobe(m) as d
+  export probe(clock) as e
+```
+
+Exporting to a field within a bundle or other statically known sub-element of
+an aggregate is allowed, for example:
+```firrtl
+module Foo:
+  input x : UInt
+  output y : {x: UInt, p: Probe<UInt>}
+  output z : Probe<UInt>[2]
+
+  wire p : UInt
+  p <= x
+  y.x <= p
+
+  export probe(p) as y.p
+  export probe(p) as z[0]
+  export probe(p) as z[1]
+```
+
+`RWProbe`{.firrtl} references to ports are not allowed on public-facing modules.
```

Comment:
why not?

Reply:
Great question.  In short, because of implementation details I wasn't able to sufficiently hide and this seemed lesser evil of available/considered options.  :disappointed: .  Thanks for bringing this up, we may be able to better here with some more thought!  Here's the explanation for why not:

All circuit elements that may be forced require special handling in the compiler -- both because their value cannot be reasoned about in the general case, but of interest here because of nets and the way force works.  The issue here is that to force a port without forcing up through to its instantiation-side wires/net (and perhaps further up an output port connected to it, etc.) we need to emit this differently or insert buffer wires.  It's my understanding that while it would solve this, changing ports to. e.g., "logic" instead of (implicitly) "wire" is not desirable/problematic, so the traditional solution has been going through all instantiations of modules with force-able ports and adding extra wires to distance uses from the forced net.  This is what we did with Signal Driving, after multiple rounds of painful debugging leading to re-discovering this behavior in a few different forms which I mention because the conclusion I gathered from that is that this is not how anyone trying to force things seemed to want or expected it to work.  So the hope here is to let things be Verilog-force'able while working the way it's expected they do.

Example of force+port interaction behaving unexpectedly: https://edaplayground.com/x/k_9r .  One solution w/buffer wires (here, moving the force to the buffer wire inside the module) is here:  https://edaplayground.com/x/AruJ .   I discounted the move-the-force-to-wire-inside here for concern that users would appreciate having the force apply to a wire on the inside, if they load up the module in their waveform viewer, and because this would need to work different if doing this from the instantiation side.

With all of this in mind, my concern is specifically that RWProbe's of ports on public-facing modules do not encode in the module's signature the fact that the port may be a force target.  As a result, any users of the module elsewhere (/ compilation of designs using these public modules, or raw verilog connecting to such) will not be aware of the need for distancing their nets from the port (adding buffer wires).

We could also require all public modules to be used conservatively in this way, as-if every port may be forced re:nets.  Within a design we can at least determine if a `RWProbe` is created or not and handle it accordingly.  This would work for FIRRTL-only scenario, but not the raw Verilog.

Anyway, hopefully this can be improved upon but the idea was for now to just disallow it and anyone wanting to force ports of a public module they instantiate can do so w/a RWProbe to its driver on the instantiation side.

Reply:
thanks for the detailed explanation!